framework/src/vectorpostprocessors/VectorPostprocessorData.C

```diff
@@ -86,22 +110,53 @@ VectorPostprocessorData::getVectorPostprocessorHelper(const VectorPostprocessorN
 bool
 VectorPostprocessorData::hasVectors(const std::string & vpp_name) const
 {
-  return _values.find(vpp_name) != _values.end();
+  auto it_pair = _vpp_data.find(vpp_name);
+  if (it_pair != _vpp_data.end())
+    return !it_pair->second._values.empty();
+
+  return false;
 }
 
 const std::vector<std::pair<std::string, VectorPostprocessorData::VectorPostprocessorState>> &
 VectorPostprocessorData::vectors(const std::string & vpp_name) const
 {
-  auto vec_pair = _values.find(vpp_name);
-  mooseAssert(vec_pair != _values.end(), "No vectors found for vpp_name: " << vpp_name);
+  auto it_pair = _vpp_data.find(vpp_name);
+  mooseAssert(it_pair != _vpp_data.end(), "No vectors found for vpp_name: " << vpp_name);
 
-  return vec_pair->second;
+  return it_pair->second._values;
 }
 
 void
 VectorPostprocessorData::copyValuesBack()
 {
-  for (const auto & it : _values)
-    for (const auto & vec_it : it.second)
-      vec_it.second.old->swap(*vec_it.second.current);
+  /**
+   * Depending on the needs of this VPP, we'll perform
+   * different actions on the vectors as time is advanced.
+   *
+   *   Need Old | Preserve History | Action
+   *  --------------------------------------
+   *    False   |      False       |  None
+   *    False   |      True        |  None
+   *    True    |      False       |  Swap
+   *    True    |      True        |  Copy
+   */
+  for (const auto & vec_pair : _vpp_data)
+  {
+    bool needs_old = vec_pair.second._needs_old;
+    bool preserve_history = vec_pair.second._contains_complete_history;
+
+    if (!needs_old)
+      continue;
+
+    for (const auto & vec_it : vec_pair.second._values)
+      if (preserve_history)
+        *vec_it.second.old = *vec_it.second.current;
+      else
```

Comment:
Wait - is this `else` actually inside that `for`?

If it is - I didn't know that was valid C++.  I would have thought you needed curly braces here...